dbms/src/Flash/Disaggregated/WNFetchPagesStreamWriter.cpp

```diff
@@ -34,56 +36,81 @@ namespace DB
+    UInt64 packet_limit_size)
 {
     return std::unique_ptr<WNFetchPagesStreamWriter>(new WNFetchPagesStreamWriter(
         task.seg_task,
         task.column_defines,
         task.output_field_types,
-        read_page_ids));
+        read_page_ids,
+        packet_limit_size));
 }
 
-disaggregated::PagesPacket WNFetchPagesStreamWriter::nextPacket()
+std::pair<DM::RemotePb::RemotePage, size_t> WNFetchPagesStreamWriter::getPersistedRemotePage(UInt64 page_id)
 {
-    // TODO: the returned rows should respect max_rows_per_chunk
+    auto page = seg_task->read_snapshot->delta->getPersistedFileSetSnapshot()->getDataProvider()->readTinyData(page_id);
+    DM::RemotePb::RemotePage remote_page;
+    remote_page.set_page_id(page_id);
+    remote_page.mutable_data()->assign(page.data.begin(), page.data.end());
+    const auto field_sizes = PageUtil::getFieldSizes(page.field_offsets, page.data.size());
+    for (const auto field_sz : field_sizes)
+    {
+        remote_page.add_field_sizes(field_sz);
+    }
+    return {remote_page, page.data.size()};
+}
 
+void WNFetchPagesStreamWriter::pipeTo(SyncPagePacketWriter * sync_writer)
+{
     disaggregated::PagesPacket packet;
-
-    // read page data by page_ids
-    size_t total_pages_data_size = 0;
-    auto persisted_cf = seg_task->read_snapshot->delta->getPersistedFileSetSnapshot();
+    MemTrackerWrapper packet_mem_tracker_wrapper(fetch_pages_mem_tracker.get());
+    UInt64 total_pages_data_size = 0;
+    UInt64 packet_count = 0;
+    UInt64 pending_pages_data_size = 0;
+    UInt64 read_page_ns = 0;
+    UInt64 send_page_ns = 0;
     for (const auto page_id : read_page_ids)
     {
-        auto page = persisted_cf->getDataProvider()->readTinyData(page_id);
-        DM::RemotePb::RemotePage remote_page;
-        remote_page.set_page_id(page_id);
-        remote_page.mutable_data()->assign(page.data.begin(), page.data.end());
-        const auto field_sizes = PageUtil::getFieldSizes(page.field_offsets, page.data.size());
-        for (const auto field_sz : field_sizes)
+        Stopwatch sw_packet;
+        auto [remote_page, page_size] = getPersistedRemotePage(page_id);
+        total_pages_data_size += page_size;
+        pending_pages_data_size += page_size;
+        packet.mutable_pages()->Add(remote_page.SerializeAsString());
+        packet_mem_tracker_wrapper.alloc(page_size);
+        read_page_ns += sw_packet.elapsedFromLastTime();
+
+        if (pending_pages_data_size > packet_limit_size)
         {
-            remote_page.add_field_sizes(field_sz);
+            ++packet_count;
+            sync_writer->Write(packet);
+            send_page_ns += sw_packet.elapsedFromLastTime();
+            pending_pages_data_size = 0;
+            packet.clear_pages(); // Only set pages field before.
+            packet_mem_tracker_wrapper.freeAll();
         }
-        total_pages_data_size += page.data.size();
-        packet.mutable_pages()->Add(remote_page.SerializeAsString());
+    }
```

Comment:
How about doing it in the following way?

```cpp
Stopwatch sw_packet;
for (size_t i = 0; i < read_page_ids.size(); ++i) 
{
    const auto & page_id = read_page_ids[i];
    auto [remote_page, page_size] = getPersistedRemotePage(page_id);
    read_page_ns += sw_packet.elapsedFromLastTime();
    if (pending_pages_data_size + page_size > packet_limit_size || i == read_page_ids.size() - 1)
    {
        ++packet_count;
        sync_writer->Write(packet);
        send_page_ns += sw_packet.elapsedFromLastTime();
        pending_pages_data_size = 0;
        packet.clear_pages(); // Only set pages field before.
        packet_mem_tracker_wrapper.freeAll();
    }
    total_pages_data_size += page_size;
    pending_pages_data_size += page_size;
    packet.mutable_pages()->Add(remote_page.SerializeAsString());
    packet_mem_tracker_wrapper.alloc(page_size);
    read_page_ns += sw_packet.elapsedFromLastTime();
}
```

To make sure packet size always `<= packet_limit_size`

Reply:
make sense